### PR TITLE
Disable 'Rename File' if we're in a CPS project.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
**Customer scenario**

Customer performs the "rename file to match type name" refactoring in a .net core project.

**Bugs this fixes:** 

https://github.com/dotnet/roslyn/issues/16546

**Workarounds, if any**

None.

**Risk**

Low.

**Performance impact**

Low.

**Is this a regression from a previous update?**

No.  This is a new feature.

**Root cause analysis:**

'Rename file to match type' works by deleting the old file and adding in the new file with the right name.  However, in .net core projects this ends up not working.  That's because the CPS project system that backs .net core projects doesn't actually provide synchronous file-remove/add (because everything is Async in that world).  Because of this, after roslyn deletes/adds the file, the file is actually not where we expect it to be.  This ends up causing machinery in Roslyn to end up crashing because this invariant is not preserved.

Currently, it is thought to be too costly to come up with proper async machinery for CPS and Roslny in the RTM timeframe.  So we're just disabling the feature in this scenario for now.

**How was the bug found?**

Several customer reports (internal and external).

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @davkean @jasonmalinowski @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
If we take this, i will open a bug to track making this feature work properly on .net core projects.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 1]
```diff
@@ -198,6 +198,21 @@ internal override bool CanChangeActiveContextDocument
             }
         }
 
+        internal override bool CanRenameFilesDuringCodeActions(CodeAnalysis.Project project)
+        {
+            if (this.TryGetHierarchy(project.Id, out var hierarchy))
+            {
+                // Currently renaming files in CPS projects (i.e. .Net Core) doesn't work proprey.
```
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: sp: `properly`.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 2]
```diff
@@ -198,6 +198,21 @@ internal override bool CanChangeActiveContextDocument
...
+                // This is because the remove/add of the documents in CPS is not synchronous
+                // (despite the DTE interfaces being synchronous).  So Roslyn calls the methods
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: Please link to the tracking bug.

File: src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.MoveTypeCodeAction.cs [hunk 3]
```diff
@@ -69,6 +69,19 @@ private Editor GetEditor(CancellationToken cancellationToken)
...
                 throw ExceptionUtilities.Unreachable;
             }
+
+            internal override bool PerformFinalApplicabilityCheck => true;
+
+            internal override bool IsApplicable(Workspace workspace)
+            {
+                switch (_operationKind)
+                {
+                    case OperationKind.RenameFile:
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.MoveTypeCodeAction.cs]: Just use an if...?

**Comment by Pilchie** [comment]
Tagging @MattGertz for approval.  Also, I'd love a review from @davkean.

**Comment by MattGertz** [comment]
The template above isn't ready, I think... :-)

**Comment by CyrusNajmabadi** [comment]
Tagging @MattGal @natidea

**Comment by MattGal** [comment]
@CyrusNajmabadi  I believe I have been mistakenly tagged, please let me know if not.

**Comment by Pilchie** [comment]
Yes, mean @MattGertz

**Pull request merged**